# Working with Numbers

At its core, Python is a calculator. For example, we could use Python to solve $8 + 4$:

In [1]:
print( 8 + 4 )

12


We can also solve $8 - 4$:

In [2]:
print( 8 - 4 )

4


We can solve $8 \times 4$ using an *asterisk* (**\***):

In [3]:
print( 8 * 4 )

32


And we can solve $8 \div 4$ using a *forward slash* (**/**):

In [4]:
print( 8 / 4 )

2.0


To solve $8^4$, we use two asterisks (**\*\***):

In [5]:
print( 8 ** 4)

4096


### Sidenote: Exponents

$8^4$ is pronounced *"$8$ to the power of $4$"*. and is the same thing as $8 \times 8 \times 8 \times 8$. Basically, you have $4$ copies of the number $8$, and you are multiplying them all together. This operation is called *exponentiation*. $8$ is the *base*, and $4$ is the *exponent*. Here are some more examples:

$2^2 = 2 \times 2 = 4$

$3^3 = 3 \times 3 \times 3 = 27$

$5^6 = 5 \times 5 \times 5 \times 5 \times 5  \times 5 = 15625$

You should also know that any number to the power of $1$ is itself, and any number to the power of $0$ is $1$.

$1000000^1 = 1000000$

$1000000^0 = 1$

## Excercises
Use Python to solve the following math problems:

1\. $256 + 957$

2\. $374 - 891$

3\. $831 \times 112$

4\. $334 \div 42$

5\. $32 ^ 6$

# Order of Operations

What is the solution to $4 + 12 \div 4$? Let's use Python to find out:

In [6]:
print( 4 + 12 / 4 )

7.0


The answer is $7$. You might be wondering why the answer isn't $4$. After all, $4 + 12 = 16$ and $16 \div 4 = 4$. But this is not the correct way to solve the problem because of the *order of operations*.

According to the order of operations, you must solve problems using the following steps:

1. First solve anything inside of parentheses, starting with the innermost parentheses and working outwards.
2. Second, solve all exponentiation from left to right.
3. Third, solve all multiplication and division from left to right.
4. Finally, solve all addition and subtraction from left to right.

Just remember "PEMDAS" for Parentheses (P), Exponents (E), Multiplication and Division (MD), and Addition and Subtraction (AS).

Let's look at an example:

$(9 - 5 \div (8 - 3) \times 2 + 6) + 3^2$

First, we need to solve what's in the parentheses. There are two sets of parentheses in this problem. We need to start with the innermost ones, which contain $8 - 3$. The answer to $8 - 3$ is $5$, so we can rewrite the problem as

$(9 - 5 \div 5 \times 2 + 6) + 3^2$

Now we have to solve what's in the remaining set of parentheses. The order of operations tells us that we need to first solve the multiplication and division problems from left to right, so first we replace $5 \div 5$ with $1$.

$(9 - 1 \times 2 + 6) + 3^2$

Then we replace $1 \times 2$ with $2$.

$(9 - 2 + 6) + 3^2$

Now we solve addition and subtraction problems from left to right, so we replace $9 - 2$ with $7$

$(7 + 6) + 3^2$

And we replace $7 + 6$ with $13$

$13 + 3^2$

Now we are just left with an addition and an exponentiation. First we solve $3^2$, which is 9

$13 + 9$

Finally, $13 + 9$ is $22$, and that is our final answer. We can check our work using Python:

In [7]:
print( (9 - 5 / (8 - 3) * 2 + 6) + 3**2 )

22.0


## Excercises

Solve the following math problems using pencil and paper. Then use Python to check your work.

1\. $5^2 + 3$

2\. $3 + 6 \times (5 + 4) \div 3 - 7$

3\. $150 \div (6 + 3 \times 8) - 5$

# Variables

Let's consider the following problem:

*Yoshi walks through the forest looking for fruit. He finds $5$ apples, $3$ oranges, $2$ bananas, and $8$ strawberries. How many pieces of fruit did Yoshi find in total?*

To solve this problem, we just need to add up all the numbers of apples, oranges, bananas, and strawberries. This is very easy to do in Python:

In [8]:
print( 5 + 3 + 2 + 8 )

18


This code gives us the correct answer, but if you turned off your computer for the night and came back the next day, you probably wouldn't remember what problem it was supposed to solve. There's nothing in this code that tells you what any of the numbers mean! To help with this issue, we can give these numbers names by creating *variables*. A variable is a value that has a name attached to it. To create a variable in Python, type the name of the variable, an equals sign (**=**), and the value of the variable, like so:

In [9]:
number_of_apples = 5

We now have a variable called "number_of_apples" which corresponds to the number of apples that Yoshi found. This is called *variable assignment* If we forget what the value of this variable is, we can just run the following code:

In [10]:
print( number_of_apples )

5


We can also do math with our variable as if it were just a normal number:

In [11]:
print( number_of_apples + 1 )

6


In [12]:
print( number_of_apples * 2 )

10


In [13]:
print( (number_of_apples ** 2 + 5) / 6 )

5.0


Note that you must always refer to a variable using its **exact** name. Any typos will cause Python to give you an error:

In [14]:
print( number_of_apple )

NameError: name 'number_of_apple' is not defined

In [15]:
print( Number_Of_Apples )

NameError: name 'Number_Of_Apples' is not defined

Also note that Python will give you an error if you try to get the value of a variable that you haven't created:

In [16]:
print( number_of_peaches )

NameError: name 'number_of_peaches' is not defined

Let's make some variables corresponding to the other fruits that Yoshi found:

In [17]:
number_of_oranges = 3
number_of_bananas = 2
number_of_strawberries = 8

Now let's create one more variable that represents the total number of fruits that Yoshi found:

In [18]:
total_number_of_fruits = number_of_apples + number_of_oranges + number_of_bananas + number_of_strawberries

Notice that you can create a variable from other variables! Now to get the final answer to the problem, we can just print the value of total_number_of_fruits:

In [19]:
print( total_number_of_fruits )

18


It should be pointed out that there is nothing special about the names of the variables we created. We could have named our variables single letters like 'x', 'y', or 'z', or nonsense words like 'foo' and 'baz'. But remember that part of the reason for creating variables is to help you understand what your code is doing, so make sure you always use helpful variable names!

There are some names that Python absolutely won't allow you to use. For example, if you try to name a variable with just a number, Python will give you an error:

In [20]:
10 = 5

SyntaxError: can't assign to literal (<ipython-input-20-6b36426d02b5>, line 1)

There are two main rules to follow when giving variables names:
1. Variable names must always start with a letter or an underscore (_).
2. The rest of the name must consist of only letters, numbers, and underscores.

These are all acceptable variable names:
- PYTHON
- _underscore
- password1
- n00b

These are **not** acceptable variable names:
- 12345
- number of apples
- (^_^)
- number+of+apples

There are also some special *keywords* that Python will not let you use as variable names: 'False', 'None', 'True', 'and', 'as', 'assert', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', and 'yield'.

We'll talk about what some of these keywords are used for in later lessons.


## Excercise

Use Python to solve the following problem. Make sure you use variables with helpful names.

*Mario has $100$ coins. He spends $25$ coins on a new hat, and then he spends $47$ coins on a new pair of overalls. Then he finds $12$ coins inside a block. How many coins does Mario have now?*

# Manipulating Variables

One of the cool things about variables is that we can change them. Recall that our variable "number_of_apples" was equal to $5$.

In [21]:
print( number_of_apples )

5


We can assign it to a different number if we want:

In [22]:
number_of_apples = 7
print( number_of_apples )

7


Or we can assign it to the value of a formula:

In [23]:
number_of_apples = number_of_bananas * 2
print( number_of_apples )

4


Let's say that Yoshi starts out with five apples, and then he finds two more. We can represent this scenario in Python code like this:

In [24]:
number_of_apples = 5
number_of_apples = number_of_apples + 2
print(number_of_apples)

7


It might look strange to have the same variable on both sides of the equals sign, but here's what is going on: Python first sees "number_of_apples =", so it knows that you want to assign the variable "number_of_apples" to some sort of value, and it looks to the right of the equals sign to figure out what that value is. It sees "number_of_apples + 2". It knows that "number_of_apples" is $5$, and $5 + 2 = 7$, so the new value of "number_of_apples" is 7.

Something you should keep in mind is that when you assign variable1 equal to variable2, the value of variable1 will be equal to the value that variable2 was at the time of assignment, even if the value of variable2 changes later. For example:

In [25]:
variable2 = 10
variable1 = variable2
variable2 = 5
print(variable1)

10


## Excercises

Try to guess what the value of the variable 'x' will be after running the following Python code. Then actually run the code to check your answer.

In [26]:
x = 6
y = 3
x = x - y + 1
print(x)

In [27]:
x = 12
y = 4
x = x / y
y = 3
x = x / y
print(y)

Use Python to solve the following problem. Make sure you use variables with helpful names, and change the values of the variable when necessary.

*Yoshi walks through the forest looking for fruit. He finds $6$ apples, but he discovers that $2$ of them are rotten, so he throws them away. Then he finds $8$ oranges and $3$ strawberries. Later on, he finds $4$ strawberries and $2$ bananas. On his way home, he drops $2$ of the oranges, and $1$ of the strawberries. How many of each type of fruit does Yoshi have? How many fruits does Yoshi have total?*

# Functions

Pretend that you own a restaurant, and you want to calculate how much money you are earning. You earn money from selling food, but you also lose money from buying supplies and paying your employees. Let's say that one day your earn $\$3224.45$ from sales, and you lose $\$457.21$ from buying supplies and $\$699.53$ from paying your employees. We can calculate your *net profit* in Python like this:

In [28]:
sales = 3224.45
supplies = 457.21
employees = 699.53
profit = sales - supplies - employees
print(profit)

2067.71


OK, so you earned $\$2067.71$ that day! On the next day you earn $\$3658.11$ from sales, you lose $\$255.66$ from buying supplies, and $\$634.95$ from paying your employees. We can write some more Python code to calculate the profit:

In [29]:
sales = 3658.11
supplies = 255.66
employees = 634.95
profit = sales - supplies - employees
print(profit)

2767.5


On that day you earned  $\$2767.50$! On the next day you earn  $\$2984.70$ from sales, you lose  $\$300.04$ from buying supplies, and  $\$702.62$  from paying your employees. Again, we can write some more Python code:

In [30]:
sales = 2984.70
supplies = 300.04
employees = 702.62
profit = sales - supplies - employees
print(profit)

1982.04


So on that day you earned $\$1982.04$. It's getting kind of annoying to keep writing this code over and over again, isn't it? It's basically the same code every time, just different numbers. This is a case where you would want to create a *function*. A function allows you to run the same code over and over again without having to type it out every time. Let's look at an example of a *function definition*:

In [31]:
def calculate_profit(sales, supplies, employees):
    profit = sales - supplies - employees
    return profit

When you create a function, you have to use the keyword "def" which is short for define. Then you put the name of the function and the function's *arguments* in parentheses. The arguments are the variables that change each time you run the function. In this case, our arguments are the money earned from sales, and the money lost from supplies and employees. Then you put a colon (**:**), and underneath that you put the *body* of the function. The body is the code that you want to run over and over again. In this case, we want to calculate the net profit from sales, supplies, and employees. The *return statement* at the end of the body tells Python what you want the final answer of the function to be. In this case, we want the final answer to be the net profit. Notice that the body must be indented. The indendation tells python where the body of the function starts and ends, and it makes the code easier for you to read.

Now let's use our function. To use it, we type out the function's name, and then put the arguments in parentheses. We have to put our arguments in the same order that we put them in the function definition. So in this case we put "calculate_profit(sales, supplies, employees):

In [32]:
day1_profit = calculate_profit(3224.45, 457.21, 699.53)
day2_profit = calculate_profit(3658.11, 255.66, 634.95)
day3_profit = calculate_profit(2984.70, 300.04, 702.62)

print(day1_profit)
print(day2_profit)
print(day3_profit)

2067.71
2767.5
1982.04


We got the same answers, and we didn't have to write the same code a bunch of times! This will be very useful if we want to keep doing these same calculations every day.

### Sidenote: Comments
Earlier, it was mentioned that you can help yourself remember what your code is doing by creating helpful variable names. You can also leave helpful messages for yourself inside of your code by using *comments*. You can add a comment for yourself using the number sign (**#**) like so:

In [33]:
def calculate_profit(sales, supplies, employees):
    #This function calculates net profit from sales, supply expenses, and exmployee expenses.
    profit = sales - supplies - employees
    return profit

It's a good idea to write a comment explaining what your all functions do because you might not remember if you don't look at your code for a while. It's also a good idea to leave a comment wherever your code gets really complicated or confusing.

## Excercises

What will Python print out when the following code is run? Try to figure it out with pencil and paper first, and then actually run the code to check your work.

In [ ]:
calculate_profit(4562.00, 555.25, 705.90)

Define a function that takes two arguments, $x$ and $y$, and returns $x^2 + y^2$.

Suppose again that you own a restaurant, and there is another variable that makes you lose money: your electricity bill. Modify the function "calculate_profit" to take this variable into account.